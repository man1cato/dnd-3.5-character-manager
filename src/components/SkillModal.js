import React from 'react';
import Modal from 'react-modal';

Modal.setAppElement('#app');

const SkillModal = ({selectedSkill, handleCloseModal}) => (
    <Modal
        isOpen={!!selectedSkill}
        onRequestClose={handleCloseModal}
        contentLabel="Selected Skill"
    >
        {selectedSkill && 
            <div>
                <h4>{selectedSkill.name}</h4>
                <h5>{selectedSkill.keyAbility}</h5>
                <p>{selectedSkill.description}</p>
                {selectedSkill.notes && <p>Notes: {selectedSkill.notes}</p>}
            </div>
        }
    </Modal>

)

export default SkillModal;