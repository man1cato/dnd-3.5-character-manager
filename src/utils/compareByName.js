export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}