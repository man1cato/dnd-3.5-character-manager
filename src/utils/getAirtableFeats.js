import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset
		let feats = []

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			const featGroup = res.data.records.map((feat) =>  ({
				id: feat.id,
				name: feat.fields.Name,
				type: feat.fields.Type.join(", ") || null,
				description: feat.fields.Description || null
			}));
			
			feats = feats.concat(featGroup)
		} while (!!offset)
		
		return _.orderBy(feats, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}