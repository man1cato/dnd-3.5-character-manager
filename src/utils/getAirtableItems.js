import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset;
		let items = [];
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			const itemGroup = res.data.records.map((item) =>  ({
				id: item.id,
				name: item.fields.Name,
				value: item.fields["Value (gp)"] || '',
				weight: item.fields["Weight (lbs)"] || '',
				category: item.fields.Category || null,
				weaponType: item.fields["Weapon Type"] || null,
				encumbrance: item.fields.Encumbrance || null,
				damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
				damageS: item.fields["Damage (S)"] || null,
				damageM: item.fields["Damage (M)"] || null,
				critical: item.fields.Critical || null,
				range: item.fields.Range || null,
				armorBonus: item.fields["Armor Bonus"] || null,
				maxDexBonus: item.fields["Max DEX Bonus"] || null
			}))
			
			items = items.concat(itemGroup)
		} while (!!offset)
		
		return _.orderBy(items, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}