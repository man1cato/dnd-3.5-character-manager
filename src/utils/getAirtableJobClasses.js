import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset
		let jobClasses = []
	
		do {
			const res = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			const jobClassGroup = res.data.records.map((jobClass) =>  ({
				id: jobClass.id,
				name: jobClass.fields.Name,
				hitDie: jobClass.fields['Hit Die'],
				proficiencies: jobClass.fields['Proficiencies'] ? jobClass.fields['Proficiencies'].join(', ') : 'None',
				skills: jobClass.fields.Skills || []
			}))
			
			jobClasses = jobClasses.concat(jobClassGroup)
		} while (!!offset)
		
		return _.orderBy(jobClasses, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}