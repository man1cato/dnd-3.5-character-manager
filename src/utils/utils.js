import _ from 'lodash'
import { sizeMods } from './staticData'


export const convertTextToArray = string => _.compact(string.split(/\n/))

export const convertInputValue = value => {
	value = Number(value)
	if (value === '' || isNaN(value)) { return 0 }
	if (value === 0) { return '' }
	return value
}

export const apiObjectToArray = obj => _.sortBy(
	_.map(obj, (val, key) => ({ id: key, ...val })), 
	['name']
)


export const rollDice = (countDice = 1, dieFaces = 6) => {
	let total = 0
	for (let i = 0; i < countDice; i++) {
		total += Math.floor((Math.random() * dieFaces) + 1)
	}
	return total
}

export const calcStartingGold = jobClassName => {
	if (_.includes(['Sorcerer', 'Wizard'], jobClassName)) {
		return rollDice(3, 4) * 10
	}
	if (_.includes(['Barbarian', 'Bard'], jobClassName)) {
		return rollDice(4, 4) * 10
	}
	if (_.includes(['Fighter', 'Paladin', 'Ranger'], jobClassName)) {
		return rollDice(6, 4) * 10
	}
	if (_.includes(['Cleric',  'Rogue'], jobClassName)) {
		return rollDice(5, 4) * 10
	}
	if (jobClassName === 'Druid') {
		return rollDice(2, 4) * 10
	}
	if (jobClassName === 'Monk') {
		return rollDice(5, 4)
	}
}

export const calcAbilityMod = abilityScore => Math.floor(abilityScore/2 - 5)


export const calcSkillPoints = (jobClassName, intMod) => {
	if (_.includes(['Cleric', 'Fighter', 'Paladin', 'Sorcerer', 'Wizard'], jobClassName)) {
		return 2 + intMod
	}
	if (_.includes(['Barbarian', 'Druid', 'Monk'], jobClassName)) {
		return 4 + intMod
	}
	if (_.includes(['Bard', 'Ranger'], jobClassName)) {
		return 6 + intMod
	}
	if (jobClassName === 'Rogue') {
		return 8 + intMod
	}
}


export const calcBaseMeleeBonus = (baseAttackBonus, strScore, size) => baseAttackBonus + calcAbilityMod(strScore) + sizeMods[size].size
export const calcBaseRangedBonus = (baseAttackBonus, dexScore, size) => baseAttackBonus + calcAbilityMod(dexScore) + sizeMods[size].size
export const calcBaseGrappleBonus = (baseAttackBonus, strScore, size) => baseAttackBonus + calcAbilityMod(strScore) + sizeMods[size].grapple 

export const calcAttackBonuses = (baseAttackBonuses, characterSize, strScore, dexScore) => ({
	melee: _.map(baseAttackBonuses, bab => bab + calcAbilityMod(strScore) + sizeMods[characterSize].size),
	ranged: _.map(baseAttackBonuses, bab => bab + calcAbilityMod(dexScore) + sizeMods[characterSize].size),
	grapple: _.map(baseAttackBonuses, bab => bab + calcAbilityMod(strScore) + sizeMods[characterSize].grapple) 
})

export const calcTotalMoney = ({ pp, gp, sp, cp }) => Number((pp * 10 + gp + sp / 10 + cp / 100).toFixed(2))

export const convertMoneyToDenominations = moneyInGp => {
	const pp = Math.trunc(moneyInGp / 10) 
	const gp = Math.trunc(moneyInGp - pp * 10) 
	const sp = Math.trunc(moneyInGp * 10 - pp * 100 - gp * 10)
	const cp = Math.trunc(moneyInGp * 100 - pp * 1000 - gp * 100 - sp * 10)

	return {	pp, gp, sp, cp	}
}

export const calcItemTotalValue = (item, qty) => {
	const totalValue = item.value * qty
	return isNaN(totalValue) ? 0 : Number(totalValue.toFixed(1))
}

export const calcItemTotalWeight = (item, qty) => {
	const totalWeight = item.weight * qty
	return isNaN(totalWeight) ? 0 : Number(totalWeight.toFixed(1))
}

export const calcEquipmentTotalValue = (equipment, items) => {
	const res = _.reduce(
		_.map(equipment, item => items[item.id].value * item.qty), 
		(total, num) => total + num, 
		0
	)
	return res.toFixed(2)
}
export const calcEquipmentTotalWeight = (equipment, items) => {
	const res = _.reduce(
		_.map(equipment, item => items[item.id].weight * item.qty), 
		(total, num) => total + num, 
		0
	)
	return res.toFixed(2)
}
