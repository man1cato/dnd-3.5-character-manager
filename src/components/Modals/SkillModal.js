import React from 'react'

import DefaultModal from './DefaultModal'
import { convertTextToArray } from '../../utils/utils'


const SkillModal = ({clickedSkill, handleCloseModal}) => (
    <DefaultModal
        clickedItem={clickedSkill}
        onRequestClose={handleCloseModal}
        contentLabel="Selected Skill"
    >
        {clickedSkill && 
            <>
                <h3>{clickedSkill.name}</h3>
                <h5>{clickedSkill.keyAbility}</h5>
                {convertTextToArray(clickedSkill.description).map((paragraph, i) => 
                    <p key={i}>{paragraph}</p>
                )}
                {clickedSkill.notes && <p>Notes: {clickedSkill.notes}</p>}
            </>
        }
    </DefaultModal>
)

export default SkillModal