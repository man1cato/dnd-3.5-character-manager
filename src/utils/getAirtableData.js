import axios from 'axios'
import _ from 'lodash'
import { convertTextToArray } from './utils' 
import dotenv from 'dotenv'
dotenv.config({ path: './config/dev.env' })

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'


export const getFeats = async () => {
	try {
		let offset
		let feats = {}

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((feat) => {
				feats[feat.id] = {
					name: feat.fields.Name,
					types: feat.fields.Types || null,
					description: feat.fields.Description || null,
					prerequisites: {
						level: feat.fields['Prereq. Level'] || null,
						class: feat.fields['Prereq. Class'] || null,
						feats: feat.fields['Prereq. Feats'] || null
					} || null
				}
			})			
		} while (!!offset)
		
		return feats
	} catch (e) {
		console.error(e)
	}
}


export const getItems = async () => {
	try {
		let offset
		let items = {}
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			res.data.records.forEach((item) => {
				items[item.id] = {
					name: item.fields.Name,
					value: item.fields["Value (gp)"] || '',
					weight: item.fields["Weight (lbs)"] || '',
					category: item.fields.Category || null,
					weaponType: item.fields["Weapon Type"] || null,
					encumbrance: item.fields.Encumbrance || null,
					damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
					damageS: item.fields["Damage (S)"] || null,
					damageM: item.fields["Damage (M)"] || null,
					critical: item.fields.Critical || null,
					range: item.fields.Range || null,
					armorBonus: item.fields["Armor Bonus"] || null,
					maxDexBonus: item.fields["Max DEX Bonus"] || null
				}
			})
		} while (!!offset)
		
		return items
	} catch (e) {
		console.error(e)
	}
}


const getLevels = async (jobClassId) => {
	try {
		const levelRes = await axios.get(`${baseUrl}/Levels?filterByFormula={Class ID}="${jobClassId}"&api_key=${apiKey}`)
		
		const levelRecords = levelRes.data.records
		let levels = {}
		for (let i = 0; i < levelRecords.length; i++) {
			const levelRecord = levelRecords[i].fields 
			levels[levelRecord.Level] = {
				saves: {
					fortitude: levelRecord['Fortitude'],
					reflex: levelRecord['Reflex'],
					will: levelRecord['Will']
				},
				baseAttackBonuses: _.compact([ levelRecord['BAB 1'], levelRecord['BAB 2'], levelRecord['BAB 3'], levelRecord['BAB 4'] ]),
				xp: levelRecord['XP Required'],
				specialAbilities: !!levelRecord['Special Abilities'] ? levelRecord['Special Abilities'] : null,
				spellsPerDay:  _.compact([ 
					levelRecord['SPD 0'], levelRecord['SPD 1'], levelRecord['SPD 2'], levelRecord['SPD 3'], levelRecord['SPD 4'], 
					levelRecord['SPD 5'], levelRecord['SPD 6'], levelRecord['SPD 7'], levelRecord['SPD 8'], levelRecord['SPD 9'] 
				])
			}
		}
		return levels
	} catch (e) {
		console.error(e)
	}
}

export const getJobClasses = async () => {
	try {
		let offset
		let jobClasses = {}
	
		do {
			const classRes = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = classRes.data.offset

			const jobClassGroup = classRes.data.records
			
			for (let i = 0; i < jobClassGroup.length; i++) {
				const jobClass = jobClassGroup[i] 
				jobClasses[jobClass.id] = {
					name: jobClass.fields.Name,
					hitDie: jobClass.fields['Hit Die'],
					proficiencies: jobClass.fields['Proficiencies'] || [],
					skills: jobClass.fields.Skills || [],
					levels: await getLevels(jobClass.id)
				}
			}
		} while (!!offset)
		
		return jobClasses
	} catch (e) {
		console.error(e)
	}
}


export const getRaces = async () => {
	try {
		let offset
		let races = {}
		do {
			const res = await axios.get(`${baseUrl}/Races?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((race) =>  {
				races[race.id] = {
					name: race.fields.Name,
					size: race.fields.Size,
					speed: race.fields['Speed (ft/rnd)'],
					racialBonuses: race.fields['Racial Bonuses'] ? convertTextToArray(race.fields['Racial Bonuses']) : null,
					defaultLanguages: race.fields['Languages (Default)'],
					bonusLanguages: race.fields['Languages (Bonus)'],
					abilityMods: {
						str: race.fields['STR Mod'],
						dex: race.fields['DEX Mod'],
						con: race.fields['CON Mod'],
						int: race.fields['INT Mod'],
						wis: race.fields['WIS Mod'],
						cha: race.fields['CHA Mod']
					},
					specialAbilities: race.fields['Special Abilities'] || [],
					favoredClass:  race.fields['Favored Class'] ? race.fields['Favored Class'][0] : null,
					iconUrl: race.fields['Icon URL']
				}
			})
			
		} while (!!offset)
		
		return races
	} catch (e) {
		console.error(e)
	}
}


export const getSkills = async () => {
   let offset
   let skills = {}
   do {
		const res = await axios.get(`${baseUrl}/Skills?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
		offset = res.data.offset
		res.data.records.forEach((skill) => {
			skills[skill.id] = {
				name: skill.fields.Name,
				keyAbility: skill.fields["Key Ability"] || null,
				description: skill.fields.Description || "",
				notes: skill.fields.Notes || ""
			}
		})
   } while (!!offset)
   
   return skills
}


export const getSpecialAbilities = async () => {
   let offset
   let specialAbilities = {}
   do {
		const res = await axios.get(`${baseUrl}/Special Abilities?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
		offset = res.data.offset
		res.data.records.forEach((specialAbility) => {
			specialAbilities[specialAbility.id] = {
				name: specialAbility.fields.Name,
				type: specialAbility.fields.Type ? specialAbility.fields.Type.join(", ") : null,
				description: specialAbility.fields.Description				
			}
		})
   } while (!!offset)
   
   return specialAbilities
}


export const getSpells =  async () => {
   let offset
   let spells = {}
   do {
		const res = await axios.get(`${baseUrl}/Spells?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '')) 
		offset = res.data.offset
		res.data.records.forEach((spell) => {
			spells[spell.id] = {
				name: spell.fields.Name,
				description: spell.fields.Description || "",
				school: spell.fields.School || null,
				components: spell.fields.Components.join(", ") || null,
				castingTime: spell.fields["Casting Time"] || null,
				range: spell.fields.Range || null,
				effectArea: spell.fields["Target, Effect, Area"] || null,
				duration: spell.fields.Duration || null,
				savingThrow: spell.fields["Saving Throw"] || null,
				spellResistance: spell.fields["Spell Resistance"] || null
			}
		})
   } while (!!offset)

   return spells
}