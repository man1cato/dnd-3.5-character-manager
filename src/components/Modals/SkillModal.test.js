import React from 'react'
import {shallow} from 'enzyme'

import SkillModal from './SkillModal'
import { apiData } from '../../tests/utils/utils'


let props

beforeAll(async () => {
   const api = await apiData()
   const skills = api.skills
   props = {
      clickedSkill: Object.values(skills)[0]
   }
})


test('should render SkillModal', () => {
   const wrapper = shallow(<SkillModal {...props} />)
   expect(wrapper).toMatchSnapshot()
})