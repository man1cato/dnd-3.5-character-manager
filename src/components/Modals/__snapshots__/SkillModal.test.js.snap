// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SkillModal 1`] = `
<DefaultModal
  clickedItem={
    Object {
      "description": "You possess knowledge of local customs and events. ",
      "keyAbility": "INT",
      "name": "Knowledge (Local)",
      "notes": "",
    }
  }
  contentLabel="Selected Skill"
>
  <h3>
    Knowledge (Local)
  </h3>
  <h5>
    INT
  </h5>
  <p
    key="0"
  >
    You possess knowledge of local customs and events. 
  </p>
</DefaultModal>
`;
